web/vibe/web/i18n.d

```diff
@@ -221,11 +222,93 @@ template tr(CTX, string LANG)
 	}
 }
 
-package string determineLanguage(alias METHOD)(scope HTTPServerRequest req)
+/// Determines a language code from the value of a header string.
+/// Returns: The best match from the Accept-Language header for a language. `null` if there is no supported language.
+public string determineLanguageByHeader(T)(string accept_language, T allowed_languages) @safe pure @nogc
+	if (isForwardRange!T)
 {
+	import std.algorithm : splitter, countUntil;
 	import std.string : indexOf;
-	import std.array;
 
+	// TODO: verify that allowed_languages doesn't contain a mix of languages with and without extra specifier for the same lanaguage (but only if one without specifier comes before those with specifier)
+	// Implementing that feature should try to give a compile time warning and not change the behaviour of this function.
+
+	if (!accept_language.length)
+		return null;
+
+	string fallback = null;
+	foreach (accept; accept_language.splitter(",")) {
+		auto sidx = accept.indexOf(';');
+		if (sidx >= 0)
+			accept = accept[0 .. sidx];
+
+		string alang, aextra;
+		auto asep = accept.countUntil!(a => a == '_' || a == '-');
+		if (asep < 0)
+			alang = accept;
+		else {
+			alang = accept[0 .. asep];
+			aextra = accept[asep + 1 .. $];
+		}
+
+		foreach (lang; allowed_languages) {
+			string lcode, lextra;
+			sidx = lang.countUntil!(a => a == '_' || a == '-');
+			if (sidx < 0)
+				lcode = lang;
+			else {
+				lcode = lang[0 .. sidx];
+				lextra = lang[sidx + 1 .. $];
+			}
+			// request en_US == serve en_US
+			if (lcode == alang && lextra == aextra)
+				return lang;
+			// request en_* == serve en
+			if (lcode == alang && !lextra.length)
+				return lang;
+			// request en* == serve en_* && be first occurence
+			if (lcode == alang && lextra.length && !fallback.length)
```

Comment:
There is one fallback case missing here: `determineLanguageByHeader("en_US,enCA", ["en_GB"]) == "en_GB")`. This should ideally still apply, though: `determineLanguageByHeader("en_US,enCA", ["en_GB", "en"]) == "en")`.